esx_service/vsan_policy.py

```diff
@@ -43,6 +44,87 @@ def create(name, content):
     return create_policy_file(filename, content)
 
 
+def update(name, content):
+    """
+    Update the content of an existing VSAN policy in the VSAN datastore.
+    Update the policy content in each VSAN object currently using the policy. If
+    a VSAN policy of the given name does not exist return an error string.
+    Return None on success.
+    """
+    path = policy_path(name)
+    if not path:
+        return ERROR_NO_VSAN_DATASTORE
+
+    err = update_policy_file_content(path, content)
+    if err:
+        return err
+
+    return update_vsan_objects_with_policy(name, content)
+
+
+def update_policy_file_content(path, content):
+    """
+    Update the VSAN policy file content. Return an error msg or None on success.
+    """
+    try:
+        with open(path) as f:
+            existing_content = f.read()
+    except OSError as e:
+        if not os.path.isfile(path):
+            return 'Error: Policy does not exist'
+        else:
+            return 'Error opening existing policy file: {0}'.format(e)
+
+    if existing_content.strip() == content.strip():
+            return 'Error: New policy is identical to old policy. Ignoring.'
+
+    # Create a temporary file so we don't corrupt an existing policy file
+    tmpfile = '{0}.tmp'.format(path)
+    err = create_policy_file(tmpfile, content)
+    if err:
+        return err
+
+    # Do an atomic rename of the tmpfile to the real policy file
+    try:
+        os.rename(tmpfile, path)
+    except OSError:
+        return 'Failed to rename {0} to {1}. Do both files still exist?'
+
+    return None
+
+
+def update_vsan_objects_with_policy(name, content):
+    """
+    Find all VSAN objects using the policy given by `name` and update the policy
+    contents in their objects. Returns an error string containing the list of
+    volumes that failed to update, or a msg if there were no volumes to update.
+    Returns None if all volumes were updated successfully.
+
+    Note: This function assumes datastore_path exists.
+    """
+    update_count = 0
+    failed_updates = []
+    dockvols_path = vmdk_utils.get_vsan_dockvols_path()
+    for v in list_volumes_and_policies():
+        if v['policy'] == name:
+            volume_name = v['volume']
+            vmdk_path = os.path.join(dockvols_path, volume_name)
+            if vsan_info.set_policy(vmdk_path, content):
+                update_count = 1
+            else:
+                failed_updates.append(volume_name)
+
+    if update_count == 0:
+        return 'No volumes using policy {0}'.format(name)
```

Comment:
So, the way it's used is actually a success, since the cli itself just prints the return value. But I agree that this is hacky, and it also makes it hard to test. I will create a better return type for this.